include/pmacc/memory/buffers/DeviceBufferIntern.hpp

```diff
@@ -311,4 +314,25 @@ namespace pmacc
         bool useOtherMemory;
     };
 
+    /** Factory for a new heap-allocated DeviceBufferIntern buffer object that is a deep copy of the given device
+     * buffer
+     *
+     * @tparam TYPE value type
+     * @tparam DIM index dimensionality
+     *
+     * @param source source device buffer
+     */
+    template<class TYPE, unsigned DIM>
+    HINLINE std::unique_ptr<DeviceBufferIntern<TYPE, DIM>> makeDeepCopy(DeviceBuffer<TYPE, DIM>& source)
```

Comment:
Note: this is logically const, but some methods used internally are mistakenly not const. So I left as a non-const.